agent/src/main/java/com/microsoft/hydralab/agent/runner/maestro/MaestroListener.java

```diff
@@ -0,0 +1,180 @@
+
+        testRun.setTestStartTimeMillis(System.currentTimeMillis());
+        testRun.addNewTimeTag("testRunStarted", System.currentTimeMillis() - recordingStartTimeMillis);
+        testRunDeviceOrchestrator.setRunningTestName(testRunDevice, "MaestroTest.testRunStarted");
+        performanceTestListener.testRunStarted();
+        performanceTestListener.testStarted("MaestroTestCase" + index);
+        testRunDeviceOrchestrator.addGifFrameAsyncDelay(testRunDevice, agentManagementService.getScreenshotDir(), 5, logger);
+    }
+
+    private void initUnitCase(String caseName, int testSeconds) {
+        final int unitIndex = index;
+        ongoingTestUnit = new AndroidTestUnit();
+        ongoingTestUnit.setNumtests(index);
+        ongoingTestUnit.setStartTimeMillis(System.currentTimeMillis() - testSeconds * 1000);
+        ongoingTestUnit.setRelStartTimeInVideo(ongoingTestUnit.getStartTimeMillis() - testSeconds * 1000 - recordingStartTimeMillis);
+        ongoingTestUnit.setCurrentIndexNum(unitIndex);
+        ongoingTestUnit.setTestName(caseName);
+        ongoingTestUnit.setTestedClass("MaestroTest");
+        ongoingTestUnit.setDeviceTestResultId(testRun.getId());
+        ongoingTestUnit.setTestTaskId(testRun.getTestTaskId());
+
+        testRun.addNewTestUnit(ongoingTestUnit);
+        testRun.addNewTimeTag(unitIndex + ". " + ongoingTestUnit.getTitle(),
+                System.currentTimeMillis() - testSeconds * 1000 - recordingStartTimeMillis);
+        testRunDeviceOrchestrator.setRunningTestName(testRunDevice, ongoingTestUnit.getTitle());
+        testRunDeviceOrchestrator.addGifFrameAsyncDelay(testRunDevice, agentManagementService.getScreenshotDir(), 5, logger);
+        performanceTestListener.testStarted("MaestroTestCase" + index);
+    }
+
+    public void startRecording(int maxTime) {
+        logger.info("Start record screen");
+        if (!testTask.isDisableRecording()) {
+            testRunDeviceOrchestrator.startScreenRecorder(testRunDevice, testRun.getResultFolder(), maxTime <= 0 ? 30 * 60 : maxTime, logger);
+        }
+        logger.info("Start gif frames collection");
+        testRunDeviceOrchestrator.startGifEncoder(testRunDevice, testRun.getResultFolder(), pkgName + ".gif");
+        logger.info("Start logcat collection");
+        testRunDeviceOrchestrator.startLogCollector(testRunDevice, pkgName, testRun, logger);
+        testRun.setLogcatPath(agentManagementService.getTestBaseRelPathInUrl(new File(testRunDevice.getLogPath())));
+        recordingStartTimeMillis = System.currentTimeMillis();
+        final String initializing = "Initializing";
+        testRunDeviceOrchestrator.setRunningTestName(testRunDevice, initializing);
+        testRun.addNewTimeTag(initializing, 0);
+    }
+
+    public void testFailed(String caseName, int testSeconds, String error) {
+        errorLogEnter("testFailed", caseName);
+        performanceTestListener.testFailure("MaestroTestCase" + index);
+        index++;
```

Comment:
It seems that index and numTests have the same value? If so, we can remove one of them.